File: SearchSortedArrayOfUnknownSize.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
// We dont have the length of the array, finding the high is the main thing in this problem
// initially setting low to 0 and high to 1, until high <= target,  updating high to high *2
class SortedRotatedArray {

    public int search(ArrayReader reader, int target) {

         int low = 0;
         int high = 1;

         while(reader.get(high) <= target) {
             low = high;
             high = high * 2;
         }

         while(low <= high) {
            int mid = low + (high-low)/2;
            // check if mid is the target
            if(reader.get(mid) == target) {
                return mid;
            }

            if(reader.get(mid) > target) {
                high = mid -1;
            }else {
                low = mid +1;
            }

         }
         return -1;
        }

}

File: SearchSortedArrayOfUnknownSize.cs
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
//This approach exponentially increases the search range by 
//doubling the high value until reader.Get(high) is greater than or equal to the target. 
//It efficiently narrows down the potential range without needing to
// check every element. Once the target is within the range, 
//binary search is applied to find the exact index.


/**
 * // This is ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * class ArrayReader {
 *     public int Get(int index) {}
 * }
 */

class Solution {
    public int Search(ArrayReader reader, int target) {
        int low = 0;
        int high = 1;
        
        while(reader.Get(high) < target){
            low = high;
            high = 2 * high;
        }
        while(low <= high){
            int mid = low + (high - low)/2;
            int midValue = reader.Get(mid);
            if(midValue == target) {
                return mid;
            }
            else if(midValue < target){
                low = mid + 1;

            }
            else{
                high = mid - 1;
            }
        }

        return -1;
    }
}